packages/openid4vc-client/src/OpenId4VcClientService.ts

```diff
@@ -1,10 +1,164 @@
-import { injectable, W3cCredentialService } from '@aries-framework/core'
+import type { AgentContext, Logger, W3cCredentialRecord } from '@aries-framework/core'
+import type { Jwt } from '@sphereon/openid4vci-client'
+
+import {
+  AgentConfig,
+  AriesFrameworkError,
+  injectable,
+  JsonEncoder,
+  JsonTransformer,
+  W3cCredentialService,
+  W3cVerifiableCredential,
+} from '@aries-framework/core'
+import {
+  Alg,
+  AuthzFlowType,
+  CredentialRequestClientBuilder,
+  OpenID4VCIClient,
+  ProofOfPossessionBuilder,
+} from '@sphereon/openid4vci-client'
+
+import { JwsService } from '../../core/src/crypto/JwsService'
+import { didKeyToVerkey } from '../../core/src/modules/dids/helpers'
+
+export interface PreAuthorizedOptions {
+  issuerUri: string
+  kid: string
+}
 
 @injectable()
 export class OpenId4VcClientService {
+  private logger: Logger
   private w3cCredentialService: W3cCredentialService
+  private jwsService: JwsService
 
-  public constructor(w3cCredentialService: W3cCredentialService) {
+  public constructor(agentConfig: AgentConfig, w3cCredentialService: W3cCredentialService, jwsService: JwsService) {
     this.w3cCredentialService = w3cCredentialService
+    this.jwsService = jwsService
+    this.logger = agentConfig.logger
+  }
+
+  private signCallback(agentContext: AgentContext) {
+    return async (jwt: Jwt, kid: string) => {
+      // TODO should we check if the did exists here, or just let the wallet throw?
+
+      if (!jwt.header) {
+        throw new AriesFrameworkError('No header present on JWT')
+      }
+
+      if (!jwt.payload) {
+        throw new AriesFrameworkError('No payload present on JWT')
+      }
+
+      const did = kid.split('#')[0]
+
+      const verkey = didKeyToVerkey(did)
+
+      const payload = JsonEncoder.toBuffer(jwt.payload)
+
+      const jws = await this.jwsService.createJwsCompact(agentContext, {
+        verkey, // FIXME null check
+        header: jwt.header as unknown as Record<string, unknown>,
+        payload,
+        protectedHeaderOptions: {
+          alg: Alg.EdDSA,
+          kid,
+        },
+      })
+
+      return jws
+    }
+  }
+
+  private getSignCallback(agentContext: AgentContext) {
+    return {
+      signCallback: this.signCallback(agentContext),
+    }
+  }
+
+  public async preAuthorized(agentContext: AgentContext, options: PreAuthorizedOptions): Promise<W3cCredentialRecord> {
+    this.logger.debug('Running pre-authorized flow with options', options)
+
+    const client = await OpenID4VCIClient.initiateFromURI({
+      issuanceInitiationURI: options.issuerUri,
+      flowType: AuthzFlowType.PRE_AUTHORIZED_CODE_FLOW,
+      kid: options.kid,
+      alg: Alg.EdDSA,
+      // The clientId is set to a dummy value for now as we don't have
+      // one in the pre-auth flow, but it's currently required by OpenID4VCIClient.
```

Comment:
Will this be fixed in the sdk that we use?

Reply:
Assuming you're talking about the comment, yes it will!